### ISSUE TITLE
zapslog: Handler should not output groups for an empty Record

### ISSUE THREAD

**Issue by arukiidou** [uber-go/zap#1401]
- part of #1334

**Describe the bug**

`slogtest` cause errors:

```
unexpected key "H": a Handler should not output groups for an empty Record (zap/exp/zapslog/slogtest.go:165)
```

actual logs:

```
zap\exp\zapslog\logger.go:130: 2024-01-28T01:08:54.927+0900	INFO	msg	{"a": "b", "G": {"c": "d", "H": {}}}
```

**To Reproduce**
run slogtest
```go
f: func(l *slog.Logger) {
	l.With("a", "b").WithGroup("G").With("c", "d").WithGroup("H").Info("msg")
},
```

**Expected behavior**

so that:

```
zap\exp\zapslog\logger.go:130: 2024-01-28T01:08:54.927+0900	INFO	msg	{"a": "b", "G": {"c": "d"}}
```

**Additional context**
none.

**Comment by arukiidou** [comment]
https://github.com/uber-go/zap/blob/70f61bb342203a50f8192b05f7faf40a0f809a91/exp/zapslog/handler.go#L62-L67

Returns an error if empty attr

**Comment by arukiidou** [comment]
### Solution

- When `WithGroup` is called, the handler "hold" the group
 - When group has not empty, "hold" is released. following:
   - `WithGroup` called again.
   - `With` called (without `slog.Attr` is empty)
   - When `Log` called with args.
   - When `LogAttrs` called (without `slog.Attr` is empty)
 - If not, don't output the `hold` group.

**Comment by abhinav** [comment]
Resolved by #1408
